core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala

```diff
@@ -124,10 +124,16 @@ private[spark] class DiskBlockManager(blockManager: BlockManager, conf: SparkCon
     (blockId, getFile(blockId))
   }
 
+  /**
+   * Create local directories for storing block data. These directories are
+   * located inside configured local directories and won't
+   * be deleted on JVM exit when using the external shuffle service.
```

Comment:
I just read it again. You mentioned that "on Mesos the shuffle dir is not deleted, but its parent directory is". I'm confused about two things: First, if the parent directory is deleted, wouldn't everything it automatically be deleted as well?

Second, I actually don't see how the parent directory (the middle layer in your comment) is deleted. Are you referring to `DiskBlockManager#doStop()`? AFAIK that only deletes the temporary directory created inside of the parent directory (i.e. the shuffle dir), and we don't do this if external shuffle service. I can't find the code that deletes the middle layer itself.